rfcs/text/0000-metadata-integration-test.md

```diff
@@ -0,0 +1,106 @@
+1. Systems that are not metadata-schema agnostic, do not fail early and predictably (should raise an error when a new major version of the schema that its developer hasn’t validated the software against before is encountered)
+1. The agility of the metadata and ingest teams is hampered
+
+To solve these issues we need to
+
+1. Make the development environment more stable
+1. Make updates to software and schema smaller and more frequent
+1. Make failures observable
+1. Make systems downstream of ingest more agnostic to schema or fail transparently
+
+## Detailed Design
+
+Below we describe a design that will transition DCP from the current "waterfall" software/metadata integration model to a more observable and interactive integration model, informed by integration tests that display the true compatibility state of the system.
+
+Setting this plan will also better define the interaction between the metadata team and the rest of the DCP. Processes and responsibilities for resolving any integration issues that arise will be clear.
+
+### Design Alternative #1
+
+Software development proceeds in the dev branch of each component independently. Changes are merged into integration branch of each component only through PRs which cause system-wide integration tests to run. PRs are merged automatically if tests pass (but can be reverted if downstream breakage occurs). PRs may not be merged if tests do not pass.
+
+![Option 1](../images/0000-metadata-integration-test-opt1.png)
+
+*Here DSS Dev is shown as an example of other components following the same process.*
+
+Data with the new schema version should not be uploaded into the `integration` or subsequent deployment stages until the test passes and the `master` branch is promoted to `integration`.
+
+#### Pros
+
+* "Test then merge"
+* PR status will not require wranglers to know about downstream systems
+* Small non-breaking changes can be merged quickly (out-of-order) and with confidence
+* Greatly reduces probability of deployment stages past integration breaking
+* More certainty for developers on behavior of the system
+* Issues caused by metadata changes are surfaced early
+* All DCP projects are integrated in the same way
+
+#### Cons
+
+* May be difficult to revert test versions of components and metadata (not all infra deployments are easily reversible)
+* Requires "serializing" the dcp-wide integration test so that one PR can run after another
+
+### Design Alternative #2 
+
+Software development proceeds in the dev branch of each component independently.
+When a metadata schema version has passed unit tests in `dev` and its committers wish to promote it, they make a PR against their integration branch.
+The PR triggers a dcp-wide test in the integration environment.
+Metadata schema changes can be merged from `dev` to `integration` only if the integration test suite passes.
```

Comment:
> When a metadata schema version has passed unit tests in `dev` and its committers wish to promote it, they make a PR against their integration branch. The PR triggers a dcp-wide test in the integration environment. Metadata schema changes can be merged from `dev` to `integration` only if the integration test suite passes.

> Data with the new schema version should not be uploaded into the `integration` or subsequent deployment stages until the test passes and the `master` branch is promoted to `integration`.

These two statements seems like they have the potential to massively delay data contributors being able to submit data to the DCP. For many reasons, this delay is not ideal. What is that rationale behind requiring all components of the DCP pass integration with the new schema before a data contributor can submit data? I'm confused as to why a contributor could not submit a dataset requiring a new field if - as an extreme example - the Data Browser fails to incorporate the new field as a facet (which could cause the integration test to fail). Any thoughts on this?

Reply:
Yes!

Unfortunately, there is some context missing here. We have been working with @tburdett on a development plan to decouple ingest from new schema support in downstream applications. Please DM me for the details. 

This brings me to the rationale... As outlined in the **Motivation** section, the rationale for these changes is to keep the development environment stable while we iterate towards having decoupling ingest/upload/storage from schema support in downstream apps. The system does _not_ currently support this.